ci: testing ASAN instrumentation of LuaJIT internal allocator

mandesero opened the pull request:
This patch is a pre-patch (refer to [1]) to add ASAN instrumentation to the LuaJIT memory allocator.

LuaJIT has two usage scenarios for ASAN:
- LuaJIT using sysmalloc
- LuaJIT using internal memory allocator

This patch introduces jobs that separate the testing of these scenarios into individual jobs. This is necessary because using the internal allocator with ASAN requires `-DLUAJIT_ENABLE_GC64=ON` (refer to [1]).

Note: this patch is required to test the ASAN instrumentation of LuaJIT internal memory allocator.

[1]: Issue https://github.com/tarantool/tarantool/issues/10231

Buristan commented on .test.mk:
```diff
@@ -90,7 +90,8 @@ CMAKE_PARAMS_ASAN = -DENABLE_WERROR=ON \
                     -DENABLE_FUZZER=ON \
-                    -DTEST_BUILD=ON
+                    -DTEST_BUILD=ON \
+		    -DLUAJIT_ENABLE_GC64=ON
```
I suggest adding `-DLUAJIT_USE_SYSMALLOC=ON` to avoid [this warning](https://github.com/tarantool/tarantool/actions/runs/9969351885/job/27546159031?pr=10232#step:8:188), unless the ASAN support for the LuaJIT internal allocator is complete.

mandesero replied:
This PR has been changed and the comment is now irrelevant. I'll let it be resolved.

Buristan reviewed (changes requested):
Hi, Maxim!
Thanks for the patch!
Please consider my comments below.

> Issue #10231

I thinks that it is `Needed for #10231`.

Also, please add a descriptive commit message.

mandesero commented:
Waiting for #10227

ligurio commented on .github/workflows/debug_asan_instrumentation_clang.yaml:
```diff
@@ -0,0 +1,119 @@
+            cd ../..
+      - name: test
+        env:
+          CC: clang-16
```
Why Clang 16? Clang 18 [^1] is available in LLVM APT repository [^2].

[^1]: https://releases.llvm.org/18.1.0/tools/clang/docs/ReleaseNotes.html
[^2]: https://apt.llvm.org/

mandesero replied:
`clang-16` is used in existing tests (`debug/release_asan_clang.yaml`), so I did not replace it with `clang-18`.

mandesero replied:
> Addressed in PR #10605.

I have rebased this PR so it now uses clang-19.

ligurio commented on .github/workflows/release_asan_instrumentation_clang.yaml:
```diff
@@ -0,0 +1,119 @@
+name: release_asan_instrumentation_clang
```
Why do we need two additional workflows instead of adding new jobs to `.github/workflows/debug_asan_clang.yml` and `.github/workflows/release_asan_clang.yml`?

ligurio commented on .test.mk:
```diff
@@ -148,6 +148,35 @@ test-debug-asan: TEST_RUN_PARAMS += --test-timeout 620 \
 
+# Release ASAN build with ASAN instrumentation of LuaJIT allocator
+
+.PHONY: test-release-asan-instrumentation
+# Copy of Release ASAN build
+test-release-asan-instrumentation: CMAKE_PARAMS = ${CMAKE_PARAMS_ASAN} \
+                                  -DLUAJIT_USE_ASAN_HARDENING=ON \
+                                  -DLUAJIT_ENABLE_GC64=ON \
+                                  -DCMAKE_BUILD_TYPE=RelWithDebInfo \
+                                  -DFIBER_STACK_SIZE=640Kb
+test-release-asan-instrumentation: TEST_RUN_ENV = ${TEST_RUN_ENV_ASAN}
+test-release-asan-instrumentation: LUAJIT_TEST_ENV = ${LUAJIT_TEST_ENV_ASAN}
+test-release-asan-instrumentation: build run-luajit-test run-test
+
+# Debug ASAN build with ASAN instrumentation of LuaJIT allocator
+
+.PHONY: test-debug-asan-instrumentation
+# Copy of Debug ASAN build
+test-debug-asan-instrumentation: CMAKE_PARAMS = ${CMAKE_PARAMS_ASAN} \
+                                -DLUAJIT_USE_ASAN_HARDENING=ON \
+                                -DLUAJIT_ENABLE_GC64=ON \
+                                -DCMAKE_BUILD_TYPE=Debug \
+                                -DFIBER_STACK_SIZE=1280Kb
+test-debug-asan-instrumentation: TEST_RUN_ENV = ${TEST_RUN_ENV_ASAN}
+test-debug-asan-instrumentation: LUAJIT_TEST_ENV = ${LUAJIT_TEST_ENV_ASAN}
+test-debug-asan-instrumentation: TEST_RUN_PARAMS += --test-timeout 620 \
+                                    --no-output-timeout 630 \
+                                    --server-start-timeout 610
+test-debug-asan-instrumentation: build run-luajit-test run-test
```
Honestly, I don't like code duplication in `.test.mk`.
I would put common parameters to a variable and use this variable in `test-debug-asan`, `test-debug-asan-instrumentation`, `test-release-asan` and `test-release-asan-instrumentation`. Is it possible?

mandesero replied:
All done, the common parameters have been moved to a variable.

ligurio commented on .test.mk:
```diff
@@ -148,6 +148,35 @@ test-debug-asan: TEST_RUN_PARAMS += --test-timeout 620 \
+
+.PHONY: test-debug-asan-instrumentation
+# Copy of Debug ASAN build
+test-debug-asan-instrumentation: CMAKE_PARAMS = ${CMAKE_PARAMS_ASAN} \
```
What is the difference for `test-debug-asan-instrumentation` and `test-debug-asan`? Both instruments build using AddressSanitizer.

ligurio reviewed (commented):
Maksim, thanks for the patch! I left a number of comments. Please take a look.

mandesero committed:
ci: add jobs for testing LuaJIT allocator & ASAN

This patch is a pre-patch (refer to [1]) to add ASAN instrumentation to
the LuaJIT memory allocator.

LuaJIT has two usage scenarios for ASAN:
- LuaJIT using sysmalloc
- LuaJIT using internal memory allocator

This patch introduces jobs that separate the testing of these scenarios
into individual jobs. This is necessary because using the embedded
allocator with ASAN requires `-DLUAJIT_ENABLE_GC64=ON` (refer to [1]).

Note: this patch is required to test the ASAN instrumentation of LuaJIT
internal memory allocator.

[1]: Issue #10231

NO_DOC=ci
NO_CHANGELOG=ci
NO_TEST=ci

mandesero commented:
After discussion, it was decided to split the use of ASAN into two scenarios:
- LuaJIT using sysmalloc: `-DLUAJIT_USE_ASAN=ON`
- LuaJIT using the internal memory allocator: `-DLUAJIT_USE_ASAN_HARDENING=ON`

This patch should add jobs that testing this scenarious.

mandesero commented:
Waiting for #10399

ligurio commented on .github/workflows/debug_asan_clang.yml:
```diff
@@ -66,6 +66,18 @@ jobs:
+        include:
+          - ALLOCATOR: sysmalloc
+            MAKE_CMD: test-debug-asan-sysmalloc
+          - ALLOCATOR: lj_alloc
```
I would replace `lj_alloc` with `dlmalloc`. Or use `internal` instead `lj_alloc` and `external` instead `sysmalloc`.

mandesero replied:
Good, replaced `lj_alloc` with `dlmalloc`.
